doc/guides/thread_local_storage.rst

```diff
@@ -0,0 +1,90 @@
+.. _thread_local_storage:
+
+Thread Local Storage (TLS)
+##########################
+
+Thread Local Storage (TLS) allows variables to be allocated on a per-thread
+basis. These variables are stored in the thread stack which means every
+thread has its own copy of these variables.
```

Comment:
Maybe a brief note that compiler-assisted TLS allows user thread access to TLS (including variables like `errno`) without making system calls.

Reply:
This is mentioned in the `CONFIG_ERRNO_IN_TLS` config below, since `errno` is the one requiring syscall without TLS.